executor/projection.go

```diff
@@ -0,0 +1,346 @@
+	if !e.prepared {
+		e.prepare(ctx)
+		e.prepared = true
+	}
+
+	output, ok := <-e.outputCh
+	if !ok {
+		return nil
+	}
+
+	err := <-output.done
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	chk.SwapColumns(output.chk)
+	e.fetcher.outputCh <- output
+	return nil
+}
+
+func (e *ProjectionExec) prepare(ctx context.Context) {
+	e.finishCh = make(chan struct{})
+	e.outputCh = make(chan *projectionOutput, e.numWorkers)
+
+	// Initialize projectionInputFetcher.
+	e.fetcher = projectionInputFetcher{
+		child:          e.children[0],
+		globalFinishCh: e.finishCh,
+		globalOutputCh: e.outputCh,
+		inputCh:        make(chan *projectionInput, e.numWorkers),
+		outputCh:       make(chan *projectionOutput, e.numWorkers),
+	}
+
+	// Initialize projectionWorker.
+	e.workers = make([]*projectionWorker, 0, e.numWorkers)
+	for i := int64(0); i < e.numWorkers; i++ {
+		e.workers = append(e.workers, &projectionWorker{
+			sctx:            e.ctx,
+			evaluatorSuit:   e.evaluatorSuit,
+			globalFinishCh:  e.finishCh,
+			inputGiveBackCh: e.fetcher.inputCh,
+			inputCh:         make(chan *projectionInput, 1),
+			outputCh:        make(chan *projectionOutput, 1),
+		})
+
+		e.fetcher.inputCh <- &projectionInput{
+			chk:          e.children[0].newChunk(),
+			targetWorker: e.workers[i],
+		}
+		e.fetcher.outputCh <- &projectionOutput{
+			chk:  e.newChunk(),
+			done: make(chan error, 1),
+		}
+	}
+
+	go e.fetcher.run(ctx)
+
+	for i := range e.workers {
+		go e.workers[i].run(ctx)
+	}
+}
+
+// Close implements the Executor Close interface.
+func (e *ProjectionExec) Close() error {
+	if e.outputCh != nil {
+		close(e.finishCh)
+		// Wait for "projectionInputFetcher" to finish and exit.
+		for range e.outputCh {
+		}
+		e.outputCh = nil
+	}
+	return errors.Trace(e.baseExecutor.Close())
+}
+
+type projectionInputFetcher struct {
+	child          Executor
+	globalFinishCh <-chan struct{}
+	globalOutputCh chan<- *projectionOutput
+
+	inputCh  chan *projectionInput
+	outputCh chan *projectionOutput
+}
+
+// run gets projectionInputFetcher's input and output resources from its
+// "inputCh" and "outputCh" channel, once the input and output resources are
+// abtained, it fetches child's result into "input.chk" and:
+//   a. Dispatches this input to the worker specified in "input.targetWorker"
+//   b. Dispatches this output to the main thread: "ProjectionExec.Next"
+//   c. Dispatches this output to the worker specified in "input.targetWorker"
+//
+// It is finished and exited once:
+//   a. There is no more input from child.
+//   b. "ProjectionExec" close the "globalFinishCh"
+func (f *projectionInputFetcher) run(ctx context.Context) {
+	defer func() {
+		close(f.globalOutputCh)
+	}()
+
+	for {
+		input := readProjectionInput(f.inputCh, f.globalFinishCh)
+		if input == nil {
+			return
+		}
+		targetWorker := input.targetWorker
+
+		output := readProjectionOutput(f.outputCh, f.globalFinishCh)
+		if output == nil {
+			return
+		}
+
+		f.globalOutputCh <- output
```

Comment:
Could this operation keep the original order?

Reply:
Yes

Reply:
This is a common pattern, dispatch to workers and keep result order at the same time:

```
type task struct {
	sync.WaitGroup
}

func main() {
	for t := range <-task {
		sendToWorker(ch, t)
		sendToKeepOrder(fifo, t)
	}

	for t := range <-fifo {
		t.wait()
	}
}

func worker(ch) {
	for t := range ch {
		t.Done()
	}
}
```

Send task to workers, at the same time, send it to a FIFO channel to keep order.
The worker will not send the result, it will just mark the done flag instead.
The collector receive from FIFO and wait for the done flag.